src/vm-vcpu-ref/src/x86_64/mptable.rs

```diff
@@ -112,229 +117,266 @@ fn mpf_intel_compute_checksum(v: &mpspec::mpf_intel) -> u8 {
+        let table_end = base_mp;
+
+        {
+            let mut mpc_table = MpcTable(mpspec::mpc_table {
+                signature: MPC_SIGNATURE,
+                // it's safe to use unchecked_offset_from because
+                // table_end > table_base
+                length: table_end.unchecked_offset_from(table_base) as u16,
+                spec: MPC_SPEC,
+                oem: MPC_OEM,
+                productid: MPC_PRODUCT_ID,
+                lapic: APIC_DEFAULT_PHYS_BASE,
+                ..Default::default()
+            });
+            checksum = checksum.wrapping_add(compute_checksum(&mpc_table.0));
+            mpc_table.0.checksum = (!checksum).wrapping_add(1) as i8;
+            mem.write_obj(mpc_table, table_base)
+                .map_err(|_| Error::WriteMpcTable)?;
+        }
 
-    Ok(())
+        Ok(())
+    }
 }
-
 #[cfg(test)]
 mod tests {
     use super::*;
-
-    use vm_memory::{self, Bytes};
-    type GuestMemoryMmap = vm_memory::GuestMemoryMmap<()>;
+    use vm_memory::{Bytes, GuestMemoryMmap};
 
     fn table_entry_size(type_: u8) -> usize {
         match u32::from(type_) {
-            mpspec::MP_PROCESSOR => mem::size_of::<MpcCpuWrapper>(),
-            mpspec::MP_BUS => mem::size_of::<MpcBusWrapper>(),
-            mpspec::MP_IOAPIC => mem::size_of::<MpcIoapicWrapper>(),
-            mpspec::MP_INTSRC => mem::size_of::<MpcIntsrcWrapper>(),
-            mpspec::MP_LINTSRC => mem::size_of::<MpcLintsrcWrapper>(),
+            mpspec::MP_PROCESSOR => mem::size_of::<MpcCpu>(),
+            mpspec::MP_BUS => mem::size_of::<MpcBus>(),
+            mpspec::MP_IOAPIC => mem::size_of::<MpcIoapic>(),
+            mpspec::MP_INTSRC => mem::size_of::<MpcIntsrc>(),
+            mpspec::MP_LINTSRC => mem::size_of::<MpcLintsrc>(),
             _ => panic!("unrecognized mpc table entry type: {}", type_),
         }
     }
 
     #[test]
     fn bounds_check() {
+        // Test mptable fits in guest memory.
         let num_cpus = 4;
-        let mem = GuestMemoryMmap::from_ranges(&[(
-            GuestAddress(MPTABLE_START),
-            compute_mp_size(num_cpus),
-        )])
-        .unwrap();
-
-        setup_mptable(&mem, num_cpus).unwrap();
-    }
-
-    #[test]
-    fn bounds_check_fails() {
-        let num_cpus = 4;
-        let mem = GuestMemoryMmap::from_ranges(&[(
-            GuestAddress(MPTABLE_START),
-            compute_mp_size(num_cpus) - 1,
-        )])
-        .unwrap();
-
-        assert!(setup_mptable(&mem, num_cpus).is_err());
+        let mptable = MpTable::new(num_cpus).unwrap();
+        let mem: GuestMemoryMmap =
+            GuestMemoryMmap::from_ranges(&[(GuestAddress(MPTABLE_START), mptable.size())]).unwrap();
+        mptable.write(&mem).unwrap();
+
+        // Test mptable does not fit in guest memory.
+        let num_cpus = 5;
+        let mptable = MpTable::new(num_cpus).unwrap();
+        assert_eq!(mptable.write(&mem).unwrap_err(), Error::NotEnoughMemory);
```

Comment:
Would it be possible to test other error conditions as well, e.g. WriteMpfIntel, WriteMpcCpu?

Reply:
I don't think it's possible to hit those errors because we're checking at the beginning if the full table is going to fit in memory, and if it doesn't that we exit immediately.